QModManager/DataStructures/SortedTree.cs

```diff
@@ -0,0 +1,217 @@
+﻿namespace QModManager.DataStructures
+{
+    using System;
+    using System.Collections.Generic;
+
+    internal class SortedTree<IdType, DataType>
+        where IdType : IEquatable<IdType>, IComparable<IdType>
+        where DataType : ISortable<IdType>
+    {
+        private SortedTreeNode<IdType, DataType> Root;
+        private readonly IDictionary<IdType, SortedTreeNode<IdType, DataType>> SortedElements;
+        private readonly ICollection<IdType> KnownKeys = new HashSet<IdType>();
+        private List<IdType> erroredIds = new List<IdType>();
+
+        internal int NodesInError;
+        internal int NodeCount => SortedElements.Count - NodesInError;
+
+        public SortedTree()
+        {
+            SortedElements = new Dictionary<IdType, SortedTreeNode<IdType, DataType>>();
+        }
+
+        public SortedTree(int capacity)
+        {
+            SortedElements = new Dictionary<IdType, SortedTreeNode<IdType, DataType>>(capacity);
+        }
+
+        public SortResults Add(DataType data)
+        {
+            if (IsDuplicateId(data.Id))
+            {
+                return SortResults.DuplicateId;
+            }
+
+            KnownKeys.Add(data.Id);
+
+            var entity = new SortedTreeNode<IdType, DataType>(data.Id, data, this);
+
+            if (Root == null)
+            {
+                Root = entity;
+                SortedElements.Add(entity.Id, entity);
+                return SortResults.SortAfter;
+            }
+
+            SortResults sortResult = Root.Sort(entity);
+
+            switch (sortResult)
+            {
+                case SortResults.SortBefore:
+                case SortResults.SortAfter:
+                    SortedElements.Add(entity.Id, entity);
+                    break;
+                default:
+                    NodesInError++;
+                    break;
+            }
+
+            return sortResult;
+        }
+
+        public List<IdType> GetErrors()
+        {
+            return erroredIds;
```

Comment:
Maybe this should be an `out` parameter on the `CreateFlatIndexList` and `CreateFlatValueList` methods.